// @obsidize/tar-browserify/header/ustar/ustar-header-link-indicator-type.js
/**
 * Sector type flag values taken from here:
 * (see "Type flag field" in wiki)
 *
 * https://en.wikipedia.org/wiki/Tar_(computing)
 *
 * Special notes from the wiki:
 * 'A'–'Z' - Vendor specific extensions (POSIX.1-1988)
 * All other values	- Reserved for future standardization
 */
export var UstarHeaderLinkIndicatorType;
(function (UstarHeaderLinkIndicatorType) {
    /**
     * Special local indicator for this module to indicate a parse failure
     */
    UstarHeaderLinkIndicatorType["UNKNOWN"] = "UNKNOWN";
    UstarHeaderLinkIndicatorType["NORMAL_FILE"] = "0";
    UstarHeaderLinkIndicatorType["NORMAL_FILE_ALT1"] = "\0";
    UstarHeaderLinkIndicatorType["NORMAL_FILE_ALT2"] = "";
    UstarHeaderLinkIndicatorType["HARD_LINK"] = "1";
    UstarHeaderLinkIndicatorType["SYMBOLIC_LINK"] = "2";
    UstarHeaderLinkIndicatorType["CHARACTER_SPECIAL"] = "3";
    UstarHeaderLinkIndicatorType["BLOCK_SPECIAL"] = "4";
    UstarHeaderLinkIndicatorType["DIRECTORY"] = "5";
    UstarHeaderLinkIndicatorType["FIFO"] = "6";
    UstarHeaderLinkIndicatorType["CONTIGUOUS_FILE"] = "7";
    /**
     * Global extended header with meta data (POSIX.1-2001)
     */
    UstarHeaderLinkIndicatorType["GLOBAL_EXTENDED_HEADER"] = "g";
    /**
     * Extended header with meta data for the next file in the archive (POSIX.1-2001)
     */
    UstarHeaderLinkIndicatorType["LOCAL_EXTENDED_HEADER"] = "x";
})(UstarHeaderLinkIndicatorType || (UstarHeaderLinkIndicatorType = {}));
